### ISSUE TITLE
[Linux] Support org.freedesktop.appearance.color-scheme to retreive system dark mode

### ISSUE THREAD

**Issue by vimpostor** [wez/wezterm#2258]
The current dark mode detection in wezterm for Linux is unreliable. Luckily, there is now a desktop environment independent way to query the system dark mode: The `org.freedesktop.appearance.color-scheme` key in the XDG desktop portal.

This works both on X11 and Wayland and it is also possible to be notified during runtime as soon as the dark mode preference changes.
Example desktop environments that have support for this key are:
- [KDE Plasma](https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/merge_requests/49)
- Gnome
- All other WMs if [darkman](https://gitlab.com/WhyNotHugo/darkman) or [color-scheme-simulator](https://gitlab.gnome.org/exalm/color-scheme-simulator) is installed

It would be great if wezterm could support reading this key to get the preferred appearance and also to react to runtime changes of the key. The advantage (next to being independent of DE) is that one would not need to poll for dark mode changes but would immediately get an interrupt once the system dark mode changes.

Related issue: #806

### More Info
- [Gnome blog post](https://blogs.gnome.org/alexm/2021/10/04/dark-style-preference/)
- [XDG Desktop Portal spec](https://github.com/flatpak/xdg-desktop-portal/blob/d7a304a00697d7d608821253cd013f3b97ac0fb6/data/org.freedesktop.impl.portal.Settings.xml#L33-L45)

### Example apps and implementations that already support this key
- [KeepassXC](https://github.com/keepassxreboot/keepassxc/pull/7422)
- [Telegram Desktop](https://github.com/telegramdesktop/tdesktop/pull/24242)
- [Qt](https://codereview.qt-project.org/c/qt/qtbase/+/396655)
- [vim](https://github.com/vimpostor/vim-lumen)

**Comment by wez** [comment]
Thanks for this suggestion! I've pushed a first-pass implementation of this: reading the appearance will read from this interface.
I still need to hook up subscribing to the settings-changed signal.

**Comment by wez** [comment]
Change notifications are now also hooked up!

This should be fixed now in `main`.

It typically takes about an hour before fixes are available as nightly builds for all platforms.  Linux builds are the fastest to build and are often available within about 20 minutes.  Windows and macOS builds take a bit longer.

Please take a few moments to try out the fix and let me know how that works out. You can find the nightly downloads for your system in the [wezterm installation docs](https://wezfurlong.org/wezterm/installation.html).

If you prefer to use packages provided by your distribution or package manager of choice and don't want to replace that with a nightly download, keep in mind that you can download portable packages (eg: a `.dmg` file on macOS, a `.zip` file on Windows and an `.AppImage` file on Linux) that can be run without permanently installing or replacing an existing package, and can then simply be deleted once you no longer need them. 

If you are eager and can [build from source](https://wezfurlong.org/wezterm/install/source.html) then you may be able to try this out more quickly.

**Comment by vimpostor** [comment]
Wow thanks, you are an absolute legend and it works beautifully!

https://user-images.githubusercontent.com/21310755/178947715-148b4c3a-a24c-413b-9480-db9585c9cbf8.mp4

I think this makes wezterm the very first terminal that supports this setting out of the box. :rocket:

**Comment by vimpostor** [comment]
Right now wezterm flickers very shortly on startup: It starts with whatever colorscheme is set in `color_scheme` in the config and then receives the `window-config-reloaded` a few ms later, setting the correct `color_scheme`. This causes a short flicker if the new color_scheme is a different one.

Is there any way to fix this or to already set the `color_scheme` based on appearance on startup?

I tried the following, but there is no `window` defined in the config return obviously. Do you know how to get the appearance setting in the return statement?

```wezterm
function scheme_for_appearance(appearance)
	if appearance:find("Dark") then
		return "Breeze"
	else
		return "PencilLight"
	end
end

wezterm.on("window-config-reloaded", function(window, pane)
	local overrides = window:get_config_overrides() or {}
	local scheme = scheme_for_appearance(window:get_appearance())
	if overrides.color_scheme ~= scheme then
		overrides.color_scheme = scheme
		window:set_config_overrides(overrides)
	end
end)

return {
-- HERE: Unfortunately window is not defined here and we can't get appearance
	color_scheme = scheme_for_appearance(window.appearance),
}
```

**Comment by wez** [comment]
Oh yeah, that's annoying.

The appearance stuff was only tied to a window method because we didn't have a way to reference GUI functions outside of a gui-initiated event callbacks until recently.  Now that we have `wezterm.gui` as an available lua module, I've added the function there.  That means that the config to react to dark mode changes is now a lot simpler:

```lua
local wezterm = require 'wezterm'

function scheme_for_appearance(appearance)
  if appearance:find("Dark") then
    return "Builtin Solarized Dark"
  else
    return "Builtin Solarized Light"
  end
end

return {
  color_scheme = scheme_for_appearance(wezterm.gui.get_appearance()),
}
```

I just pushed support for this, so give it 20+ minutes to show up in a nightly build!
